### ISSUE TITLE
REALM

### ISSUE THREAD

**Issue by aced125** [huggingface/transformers#3497]
# 🌟 New model addition

REALM is from some of the authors of BERT (I like to think of it as the next BERT :) ) that have found a way to incorporate world knowledge (from Wikipedia) into the model.

They do this by having the concept of a retriever module that retrieves information from wikipedia articles.

## Open source status

Code not released at the moment but will probably be released by Google soon I'd imagine.

https://arxiv.org/abs/2002.08909

**Comment by AlecS12** [comment]
Code is released for the preceding paper: Latent Retrieval for Weakly Supervised Open Domain Question Answering" https://www.aclweb.org/anthology/P19-1612/
https://github.com/google-research/language/tree/master/language/orqa

**Comment by antoniolanza1996** [comment]
I am also interested on it. Any news? Thanks

**Comment by alexvaca0** [comment]
I'm also interested on this, are you planning to include this model in the library??

**Comment by antoniolanza1996** [comment]
Finally the REALM code has been released here https://github.com/google-research/language/tree/master/language/realm.
I think that this issue should be re-opened. @aced125  are you able to re-open it?

**Comment by Utomo88** [comment]
I hope somebody can reopen this and make it happen

**Comment by krrishdholakia** [comment]
Any updates on this ?

**Comment by shamanez** [comment]
Any update?

**Comment by shamanez** [comment]
I think this could be implemented easily with RAG by making sure that we
both finetune doc encoder and question encoder of the retriever model. This
might be very useful for us.

**Comment by shamanez** [comment]
@LysandreJik  

I have implemented the RAG end-to-end training of question encoder, doc encoder (context encoder), and the generator. It is not exactly REALM but it consists of embedding re-computing and indexing with an updated context encoder within the training loop. It uses RAY. 

Can you please give it a look?

**Comment by OctoberChang** [comment]
Any Update on this?

**Comment by shamanez** [comment]
@OctoberChang  I did extend the RAG in a way that can be used to experiment REALM stuff. 

https://paperswithcode.com/paper/fine-tune-the-entire-rag-architecture

**Comment by OctoberChang** [comment]
@shamanez , thanks for the pointer to your RAG project [[link]!(https://github.com/huggingface/transformers/tree/master/examples/research_projects/rag-end2end-retriever)].
Do you have any pre-trained (not fine-tuned on downstream tasks) models with ICT or Salient span masking [[link]](https://arxiv.org/pdf/2002.08909.pdf) that can be load from Huggingface Transformers?

**Comment by shamanez** [comment]
I haven't extended in to the ICT. Sorry.
